File: backend/app/routers/schedule.py
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas import (
    DayScheduleResponse,
    MonthScheduleResponse,
    ScheduleEntryCreate,
    ScheduleEntryRead,
    ScheduleEntryUpdate,
)
from app.services import schedule_service

router = APIRouter(prefix="/schedule", tags=["schedule"])


@router.get("/day", response_model=DayScheduleResponse)
def get_day_schedule(
    target_date: date = Query(..., alias="date"),
    db: Session = Depends(get_db),
):
    return schedule_service.get_day_schedule(db, target_date)


@router.get("/month", response_model=MonthScheduleResponse)
def get_month_schedule(year: int, month: int, db: Session = Depends(get_db)):
    if month < 1 or month > 12:
        raise HTTPException(status_code=400, detail="Month must be between 1 and 12")
    return schedule_service.get_month_schedule(db, year, month)


@router.post(
    "/entries",
    response_model=ScheduleEntryRead,
    status_code=status.HTTP_201_CREATED,
)
def create_schedule_entry(
    entry: ScheduleEntryCreate, db: Session = Depends(get_db)
):
    try:
        return schedule_service.create_schedule_entry(db, entry)
    except schedule_service.ScheduleTaskNotFound as exc:
        raise HTTPException(status_code=404, detail="Task not found") from exc


@router.put("/entries/{entry_id}", response_model=ScheduleEntryRead)
def update_schedule_entry(
    entry_id: int,
    entry: ScheduleEntryUpdate,
    db: Session = Depends(get_db),
):
    updated = schedule_service.update_schedule_entry(db, entry_id, entry)
    if updated is None:
        raise HTTPException(status_code=404, detail="Schedule entry not found")
    return updated


@router.delete("/entries/{entry_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_schedule_entry(entry_id: int, db: Session = Depends(get_db)):
    if not schedule_service.delete_schedule_entry(db, entry_id):
        raise HTTPException(status_code=404, detail="Schedule entry not found")

File: frontend/src/api/schedule.js
const BASE = '/schedule'

async function parse(res) {
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`请求失败 (${res.status}) ${text}`)
  }
  if (res.status === 204) return null
  return res.json()
}

export function getDaySchedule(date) {
  return fetch(`${BASE}/day?date=${encodeURIComponent(date)}`).then(parse)
}

export function getMonthSchedule({ year, month }) {
  return fetch(
    `${BASE}/month?year=${encodeURIComponent(year)}&month=${encodeURIComponent(month)}`
  ).then(parse)
}

export function createScheduleEntry(payload) {
  return fetch(`${BASE}/entries`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  }).then(parse)
}

export function updateScheduleEntry(id, payload) {
  return fetch(`${BASE}/entries/${id}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  }).then(parse)
}

export function deleteScheduleEntry(id) {
  return fetch(`${BASE}/entries/${id}`, { method: 'DELETE' }).then(parse)
}
